// @datadog/browser-rum-core/cjs/domain/action/trackAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACTION_CONTEXT_TIME_OUT_DELAY = void 0;
exports.startActionTracker = startActionTracker;
const browser_core_1 = require("@datadog/browser-core");
const trackEventCounts_1 = require("../trackEventCounts");
exports.ACTION_CONTEXT_TIME_OUT_DELAY = 5 * browser_core_1.ONE_MINUTE; // arbitrary
function startActionTracker(lifeCycle) {
    const history = (0, browser_core_1.createValueHistory)({ expireDelay: exports.ACTION_CONTEXT_TIME_OUT_DELAY });
    const activeEventCountSubscriptions = new Set();
    const sessionRenewalSubscription = lifeCycle.subscribe(10 /* LifeCycleEventType.SESSION_RENEWED */, () => {
        history.reset();
        activeEventCountSubscriptions.forEach((subscription) => subscription.stop());
        activeEventCountSubscriptions.clear();
    });
    function createTrackedAction(startClocks) {
        const id = (0, browser_core_1.generateUUID)();
        const historyEntry = history.add(id, startClocks.relative);
        let duration;
        const eventCountsSubscription = (0, trackEventCounts_1.trackEventCounts)({
            lifeCycle,
            isChildEvent: (event) => event.action !== undefined &&
                (Array.isArray(event.action.id) ? event.action.id.includes(id) : event.action.id === id),
        });
        activeEventCountSubscriptions.add(eventCountsSubscription);
        function cleanup() {
            eventCountsSubscription.stop();
            activeEventCountSubscriptions.delete(eventCountsSubscription);
        }
        return {
            id,
            startClocks,
            get duration() {
                return duration;
            },
            get counts() {
                return eventCountsSubscription.eventCounts;
            },
            stop(endClocks) {
                historyEntry.close(endClocks.relative);
                duration = (0, browser_core_1.elapsed)(startClocks.timeStamp, endClocks.timeStamp);
                cleanup();
            },
            discard() {
                historyEntry.remove();
                cleanup();
            },
        };
    }
    function findActionId(startTime) {
        const ids = history.findAll(startTime);
        return ids.length ? ids : undefined;
    }
    function stop() {
        sessionRenewalSubscription.unsubscribe();
        activeEventCountSubscriptions.forEach((subscription) => subscription.stop());
        activeEventCountSubscriptions.clear();
        history.reset();
        history.stop();
    }
    return { createTrackedAction, findActionId, stop };
}
