// @datadog/browser-rum/esm/domain/record/startFullSnapshots.js
import { getScrollX, getScrollY, getViewportDimension } from '@datadog/browser-rum-core';
import { ExperimentalFeature, isExperimentalFeatureEnabled, timeStampNow } from '@datadog/browser-core';
import { RecordType } from '../../types';
import { createRootInsertionCursor, serializeChangesInTransaction, serializeDocument, serializeInTransaction, serializeNodeAsChange, } from './serialization';
import { getVisualViewport } from './viewports';
export function startFullSnapshots(lifeCycle, emitRecord, emitStats, flushMutations, scope) {
    takeFullSnapshot(timeStampNow(), 0 /* SerializationKind.INITIAL_FULL_SNAPSHOT */, emitRecord, emitStats, scope);
    const { unsubscribe } = lifeCycle.subscribe(2 /* LifeCycleEventType.VIEW_CREATED */, (view) => {
        flushMutations();
        takeFullSnapshot(view.startClocks.timeStamp, 1 /* SerializationKind.SUBSEQUENT_FULL_SNAPSHOT */, emitRecord, emitStats, scope);
    });
    return {
        stop: unsubscribe,
    };
}
export function takeFullSnapshot(timestamp, kind, emitRecord, emitStats, scope) {
    const { width, height } = getViewportDimension();
    emitRecord({
        data: {
            height,
            href: window.location.href,
            width,
        },
        type: RecordType.Meta,
        timestamp,
    });
    emitRecord({
        data: {
            has_focus: document.hasFocus(),
        },
        type: RecordType.Focus,
        timestamp,
    });
    if (isExperimentalFeatureEnabled(ExperimentalFeature.USE_CHANGE_RECORDS)) {
        if (kind === 1 /* SerializationKind.SUBSEQUENT_FULL_SNAPSHOT */) {
            scope.resetIds();
        }
        serializeChangesInTransaction(kind, emitRecord, emitStats, scope, timestamp, (transaction) => {
            serializeNodeAsChange(createRootInsertionCursor(scope.nodeIds), document, scope.configuration.defaultPrivacyLevel, transaction);
        });
    }
    else {
        serializeInTransaction(kind, emitRecord, emitStats, scope, (transaction) => {
            transaction.add(serializeFullSnapshotRecord(timestamp, transaction));
        });
    }
    if (window.visualViewport) {
        emitRecord({
            data: getVisualViewport(window.visualViewport),
            type: RecordType.VisualViewport,
            timestamp,
        });
    }
}
function serializeFullSnapshotRecord(timestamp, transaction) {
    return {
        data: {
            node: serializeDocument(document, transaction),
            initialOffset: {
                left: getScrollX(),
                top: getScrollY(),
            },
        },
        type: RecordType.FullSnapshot,
        timestamp,
    };
}
